File: master_be/app/services/math_lab/registry.py
from __future__ import annotations

from dataclasses import dataclass

from app.schemas.math_lab import MathDomain, MathSemanticModel, VisualizationId


@dataclass(frozen=True, slots=True)
class VisualizationDefinition:
    id: VisualizationId
    supported_domains: frozenset[MathDomain]
    min_grade: int = 1
    max_grade: int = 9
    required_features: frozenset[str] = frozenset()

    def supports(self, model: MathSemanticModel) -> bool:
        return (
            self.min_grade <= model.grade <= self.max_grade
            and (model.domain in self.supported_domains or MathDomain.UNKNOWN in self.supported_domains)
        )


class VisualizationRegistry:
    def __init__(self, definitions: tuple[VisualizationDefinition, ...]) -> None:
        self._definitions = {definition.id: definition for definition in definitions}
        if len(self._definitions) != len(definitions):
            raise ValueError("Visualization registry contains duplicate IDs")

    def get(self, visualization_id: VisualizationId) -> VisualizationDefinition:
        try:
            return self._definitions[visualization_id]
        except KeyError as exc:
            raise KeyError(f"Visualization is not registered: {visualization_id}") from exc

    def supports(self, visualization_id: VisualizationId, model: MathSemanticModel) -> bool:
        return self.get(visualization_id).supports(model)

    @property
    def ids(self) -> tuple[VisualizationId, ...]:
        return tuple(self._definitions)


DEFAULT_VISUALIZATION_REGISTRY = VisualizationRegistry(
    (
        VisualizationDefinition(VisualizationId.UNSUPPORTED, frozenset({MathDomain.UNKNOWN}), 1, 9),
        VisualizationDefinition(VisualizationId.OBJECT_GROUP, frozenset({MathDomain.ARITHMETIC, MathDomain.NUMBER, MathDomain.FRACTION, MathDomain.PROBABILITY, MathDomain.STATISTICS}), 1, 3),
        VisualizationDefinition(VisualizationId.NUMBER_LINE, frozenset({MathDomain.NUMBER, MathDomain.ARITHMETIC, MathDomain.ALGEBRA, MathDomain.COORDINATE, MathDomain.PROBABILITY, MathDomain.UNKNOWN}), 1, 9),
        VisualizationDefinition(VisualizationId.GROUPING, frozenset({MathDomain.ARITHMETIC, MathDomain.RATIO}), 1, 6),
        VisualizationDefinition(VisualizationId.BAR_MODEL, frozenset({MathDomain.ARITHMETIC, MathDomain.NUMBER, MathDomain.FRACTION, MathDomain.RATIO, MathDomain.MEASUREMENT, MathDomain.GEOMETRY_2D, MathDomain.STATISTICS, MathDomain.ALGEBRA}), 2, 9),
        VisualizationDefinition(VisualizationId.FRACTION, frozenset({MathDomain.FRACTION, MathDomain.RATIO}), 2, 9),
        VisualizationDefinition(VisualizationId.POWER_MODEL, frozenset({MathDomain.FRACTION, MathDomain.NUMBER, MathDomain.ARITHMETIC, MathDomain.ALGEBRA}), 4, 9),
        VisualizationDefinition(VisualizationId.CLOCK, frozenset({MathDomain.TIME, MathDomain.MOTION}), 1, 7),
        VisualizationDefinition(VisualizationId.TIMELINE, frozenset({MathDomain.TIME, MathDomain.MOTION}), 3, 9),
        VisualizationDefinition(VisualizationId.MOTION_PATH, frozenset({MathDomain.MOTION}), 3, 9),
        VisualizationDefinition(VisualizationId.BALANCE, frozenset({MathDomain.ALGEBRA}), 5, 9),
        VisualizationDefinition(VisualizationId.GEOMETRY_2D, frozenset({MathDomain.GEOMETRY_2D, MathDomain.MEASUREMENT}), 1, 9),
        VisualizationDefinition(VisualizationId.COORDINATE_GRAPH, frozenset({MathDomain.COORDINATE, MathDomain.ALGEBRA, MathDomain.MOTION}), 5, 9),
        VisualizationDefinition(VisualizationId.GEOMETRY_3D, frozenset({MathDomain.GEOMETRY_3D, MathDomain.MEASUREMENT}), 1, 9),
        VisualizationDefinition(VisualizationId.PLACE_VALUE, frozenset({MathDomain.NUMBER, MathDomain.ARITHMETIC}), 1, 6),
        VisualizationDefinition(VisualizationId.COLUMN_ALGORITHM, frozenset({MathDomain.NUMBER, MathDomain.ARITHMETIC}), 1, 9),
        VisualizationDefinition(VisualizationId.UNIT_SCALE, frozenset({MathDomain.MEASUREMENT}), 2, 7),
        VisualizationDefinition(VisualizationId.DATA_CHART, frozenset({MathDomain.STATISTICS, MathDomain.PROBABILITY}), 2, 9),
        VisualizationDefinition(VisualizationId.NUMBER_COMPARE, frozenset({MathDomain.NUMBER, MathDomain.ARITHMETIC}), 1, 9),
        VisualizationDefinition(VisualizationId.PROBABILITY_SIMULATOR, frozenset({MathDomain.PROBABILITY}), 5, 9),
        VisualizationDefinition(VisualizationId.PERCENT_GRID, frozenset({MathDomain.RATIO, MathDomain.FRACTION, MathDomain.ARITHMETIC, MathDomain.ALGEBRA}), 4, 9),
        VisualizationDefinition(VisualizationId.ALGEBRA_TILES, frozenset({MathDomain.ALGEBRA}), 6, 9),
        VisualizationDefinition(VisualizationId.PART_WHOLE, frozenset({MathDomain.ARITHMETIC, MathDomain.NUMBER}), 1, 3),
        VisualizationDefinition(VisualizationId.SHAPE_PATTERN, frozenset({MathDomain.GEOMETRY_2D}), 1, 3),
        VisualizationDefinition(VisualizationId.CIRCLE_MODEL, frozenset({MathDomain.GEOMETRY_2D, MathDomain.MEASUREMENT}), 3, 9),
        VisualizationDefinition(VisualizationId.FACTOR_LATTICE, frozenset({MathDomain.NUMBER, MathDomain.ARITHMETIC}), 4, 9),
        VisualizationDefinition(VisualizationId.EXPRESSION_TREE, frozenset({MathDomain.ARITHMETIC, MathDomain.NUMBER, MathDomain.FRACTION, MathDomain.ALGEBRA, MathDomain.RATIO}), 1, 9),
        VisualizationDefinition(VisualizationId.CALENDAR, frozenset({MathDomain.TIME, MathDomain.MEASUREMENT, MathDomain.NUMBER}), 1, 9),
        VisualizationDefinition(VisualizationId.SOLUTION_SET, frozenset({MathDomain.ALGEBRA, MathDomain.NUMBER}), 6, 9),
    )
)

File: frontend/src/features/math-lab/visualizations/registry.js
import { MathLabValidationError, VISUALIZATION_IDS } from "../schemas/mathLabSchema.js";
import ClockRenderer from "../renderers/ClockRenderer.jsx";
import FractionRenderer from "../renderers/FractionRenderer.jsx";
import Geometry2DRenderer from "../renderers/Geometry2DRenderer.jsx";
import MotionPathRenderer from "../renderers/MotionPathRenderer.jsx";
import NumberLineRenderer from "../renderers/NumberLineRenderer.jsx";
import ObjectGroupRenderer from "../renderers/ObjectGroupRenderer.jsx";
import TimelineRenderer from "../renderers/TimelineRenderer.jsx";
import BarModelRenderer from "../renderers/BarModelRenderer.jsx";
import BalanceRenderer from "../renderers/BalanceRenderer.jsx";
import CoordinateGraphRenderer from "../renderers/CoordinateGraphRenderer.jsx";
import Geometry3DRenderer from "../renderers/Geometry3DRenderer.jsx";
import GroupingRenderer from "../renderers/GroupingRenderer.jsx";
import PowerModelRenderer from "../renderers/PowerModelRenderer.jsx";
import UnsupportedRenderer from "../renderers/UnsupportedRenderer.jsx";
import PlaceValueRenderer from "../renderers/PlaceValueRenderer.jsx";
import ColumnAlgorithmRenderer from "../renderers/ColumnAlgorithmRenderer.jsx";
import UnitScaleRenderer from "../renderers/UnitScaleRenderer.jsx";
import DataChartRenderer from "../renderers/DataChartRenderer.jsx";
import NumberCompareRenderer from "../renderers/NumberCompareRenderer.jsx";
import ProbabilitySimulatorRenderer from "../renderers/ProbabilitySimulatorRenderer.jsx";
import PercentGridRenderer from "../renderers/PercentGridRenderer.jsx";
import AlgebraTilesRenderer from "../renderers/AlgebraTilesRenderer.jsx";
import PartWholeRenderer from "../renderers/PartWholeRenderer.jsx";
import ShapePatternRenderer from "../renderers/ShapePatternRenderer.jsx";
import CircleModelRenderer from "../renderers/CircleModelRenderer.jsx";
import FactorLatticeRenderer from "../renderers/FactorLatticeRenderer.jsx";
import ExpressionTreeRenderer from "../renderers/ExpressionTreeRenderer.jsx";
import CalendarRenderer from "../renderers/CalendarRenderer.jsx";
import SolutionSetRenderer from "../renderers/SolutionSetRenderer.jsx";

/**
 * Which component draws which visualization.
 *
 * `MathSceneRenderer` looks up here and nowhere else, so a new visualization
 * arrives by registering it rather than by editing a switch. The scene
 * renderer still handles a missing future renderer explicitly, so a partial
 * registry change cannot make the lab go blank.
 */
const RENDERERS = {
  unsupported: UnsupportedRenderer,
  object_group: ObjectGroupRenderer,
  grouping: GroupingRenderer,
  bar_model: BarModelRenderer,
  number_line: NumberLineRenderer,
  fraction: FractionRenderer,
  power_model: PowerModelRenderer,
  clock: ClockRenderer,
  timeline: TimelineRenderer,
  motion_path: MotionPathRenderer,
  geometry_2d: Geometry2DRenderer,
  balance: BalanceRenderer,
  coordinate_graph: CoordinateGraphRenderer,
  geometry_3d: Geometry3DRenderer,
  place_value: PlaceValueRenderer,
  column_algorithm: ColumnAlgorithmRenderer,
  unit_scale: UnitScaleRenderer,
  data_chart: DataChartRenderer,
  number_compare: NumberCompareRenderer,
  probability_simulator: ProbabilitySimulatorRenderer,
  percent_grid: PercentGridRenderer,
  algebra_tiles: AlgebraTilesRenderer,
  part_whole: PartWholeRenderer,
  shape_pattern: ShapePatternRenderer,
  circle_model: CircleModelRenderer,
  factor_lattice: FactorLatticeRenderer,
  expression_tree: ExpressionTreeRenderer,
  calendar: CalendarRenderer,
  solution_set: SolutionSetRenderer,
};

/** Vietnamese names, for the inspector and the teacher's own overrides. */
const LABELS = {
  unsupported: "Chưa có mô hình chính xác",
  object_group: "Nhóm đồ vật",
  number_line: "Tia số",
  grouping: "Gộp nhóm",
  bar_model: "Sơ đồ đoạn thẳng",
  fraction: "Mô hình phân số",
  power_model: "Lũy thừa phân số",
  clock: "Đồng hồ",
  timeline: "Dòng thời gian",
  motion_path: "Hành trình",
  balance: "Cân bằng phương trình",
  geometry_2d: "Hình học phẳng",
  coordinate_graph: "Đồ thị toạ độ",
  geometry_3d: "Hình học không gian",
  place_value: "Giá trị hàng và khối cơ số 10",
  column_algorithm: "Đặt tính nhớ – mượn",
  unit_scale: "Thang đổi đơn vị",
  data_chart: "Biểu đồ dữ liệu",
  number_compare: "So sánh và làm tròn trên tia số",
  probability_simulator: "Mô phỏng xác suất thực nghiệm",
  percent_grid: "Lưới 100 ô phần trăm",
  algebra_tiles: "Tile đa thức",
  part_whole: "Sơ đồ phần–toàn bộ",
  shape_pattern: "Overlay đếm và quy luật hình",
  circle_model: "Hình tròn và diện tích",
  expression_tree: "Cây biểu thức",
  calendar: "Lịch và khoảng thời gian",
  factor_lattice: "Lưới bội và ước chung",
  solution_set: "Tập nghiệm trên trục số",
};

const definitions = VISUALIZATION_IDS.map((id) => Object.freeze({
  id,
  label: LABELS[id] || id,
  renderer: RENDERERS[id] || null,
  ready: Boolean(RENDERERS[id]),
  phase: 3,
}));

export const visualizationRegistry = new Map(definitions.map((definition) => [definition.id, definition]));

export function visualizationDefinition(id) {
  const definition = visualizationRegistry.get(id);
  if (!definition) throw new MathLabValidationError(`unregistered visualization ${id}`, "visualization");
  return definition;
}

/** The types that can actually draw something today. */
export function readyVisualizations() {
  return definitions.filter((definition) => definition.ready);
}

export function validateVisualizationDescriptor(descriptor) {
  if (!descriptor || typeof descriptor !== "object" || Array.isArray(descriptor)) {
    throw new MathLabValidationError("descriptor must be an object", "visualization");
  }
  visualizationDefinition(descriptor.type);
  if (typeof descriptor.id !== "string" || !descriptor.id) {
    throw new MathLabValidationError("descriptor id is required", "visualization.id");
  }
  return structuredClone(descriptor);
}
